Model/Api/Tax.php

```diff
@@ -0,0 +1,181 @@
+    }
+
+    /**
+     * @param array $addressData
+     * @param array $shipping_option
+     * @return TotalsInformationInterface
+     */
+    public function setAddressInformation($addressData, $shipping_option)
+    {
+        $address = $this->populateAddress($addressData);
+        $this->addressInformation->setAddress($address);
+
+        if (!$shipping_option) {
+            return ;
+        }
+
+        $selectedOption = $shipping_option['reference'];
+        list($carrierCode, $methodCode) = explode('_', $selectedOption);
+
+        $this->addressInformation->setShippingCarrierCode($carrierCode);
+        $this->addressInformation->setShippingMethodCode($methodCode);
+    }
+
+    /**
+     * @param TotalsInterface $totalsInformation
+     * @param string $currencyCode
+     * @return TaxResultInterface
+     * @throws \Exception
+     */
+    public function createTaxResult($totalsInformation, $currencyCode)
+    {
+        $majorAmount = $totalsInformation->getTaxAmount();
+        $taxAmount = CurrencyUtils::toMinor($majorAmount, $currencyCode);
+
+        /**
+         * @var TaxResultInterface $taxResult
+         */
+        $taxResult = $this->taxResultFactory->create();
+        $taxResult->setSubtotalAmount($taxAmount);
+
+        return $taxResult;
+    }
+
+    /**
+     * @param TotalsInterface $totalsInformation
+     * @param string $currencyCode
+     * @param array $shipping_option
+     * @return ShippingOptionInterface
+     * @throws \Exception
+     */
+    public function createShippingOption($totalsInformation, $currencyCode, $shipping_option)
+    {
+        $shippingOption = $this->shippingOptionFactory->create();
+        $shippingOption->setTaxAmount(CurrencyUtils::toMinor($totalsInformation->getShippingTaxAmount(), $currencyCode));
+        $shippingOption->setService(@$shipping_option['service']);
+        $shippingOption->setCost(CurrencyUtils::toMinor($totalsInformation->getShippingAmount(), $currencyCode));
+        $shippingOption->setReference(@$shipping_option['reference']);
+
+        return $shippingOption;
+    }
+
+    /**
+     * @param array $addressData
+     * @param array $shipping_option
+     * @return TaxDataInterface
+     * @throws \Exception
+     */
+    public function generateResult($addressData, $shipping_option)
+    {
+        $this->setAddressInformation($addressData, $shipping_option);
+
+        $totalsInformation = $this->totalsInformationManagement->calculate(
+            $this->quote->getId(),
+            $this->addressInformation
+        );
+
+        $currencyCode = $this->quote->getQuoteCurrencyCode();
+
+        $taxResult = $this->createTaxResult($totalsInformation, $currencyCode);
+        $shippingOption = $this->createShippingOption($totalsInformation, $currencyCode, $shipping_option);
+
+        // shipping tax is already included, don't count it twice
+        $taxResult->setSubtotalAmount($taxResult->getSubtotalAmount() - $shippingOption->getTaxAmount());
```

Comment:
@njfaries, A couple of notes.
First, the main idea behind how the Shipping and Tax split is implemented is to defer all the processing to the exact Magento methods that are used in the native checkout and REMOVE ALL THE LOGIC AROUND IT FROM THE PLUGIN. Just pass in the (cart and shipping) data. In the previous Shipping and Tax implementation we are trying to do what Magento does, here we let Magento do it.
Next, subtotal amount in the tax result object is a term defined in Bolt. Bolt expects tax broken down into tax for product (subtotal) and tax on shipping (if available, otherwise 0 if shipping is not taxed)
Finally, the taxAmount returned from Magento has the shipping tax included, so we need to deduct it somewhere.
Anyway, your assumptions were correct and I will address your question by excluding the shipping from the subtotal tax at the time subtotal tax is set, not afterwards. Will update the PR.